lib/vector-config/src/str.rs

```diff
@@ -0,0 +1,12 @@
+use crate::Configurable;
+
+/// A string-like type that can be represented in a Vector configuration.
+///
+/// This is specifically used for constraining the implementation of anything map-like as objects,
+/// which maps are represented by, can only have string-like keys.
+///
+/// If this trait is implemented for a type that is not string-like, things will probably break.
+/// Don't implement this for things that are not string-like.
+pub trait ConfigurableString: Configurable {}
```

Comment:
Is there a way to enforce that? Something like also needing to `impl` `fmt` things?

Reply:
Unfortunately not.

We can't statically determine what type of schema a `Configurable`-like thing will generate. That's why this is only a _promise_ that the resulting schema that gets generated from a type implementing `ConfigurableString` will be that of a `string`.

Reply:
Well, I should say: we could also add runtime checks when generating the schema, so that any map type can generate the schema for its key type, `K`, and inspect it to make sure it's only `string`.

Runtime errors are a little less ergonomic, though, which is why I wanted a compile-time-based approach... but there's no reason it can't simply be the first line of defense rather than the only line of defense... 🤔

I'll try to whip something up.

Reply:
Yeah, figured a helpful note in the comments to _not_ do something is good, but ofc having checking and contracts in place is better. Probably not worth a lot of time investment?

Reply:
I added runtime checks for this in 1c701faec, plus a unit test for every map type we support `Configurable` for that ensures we panic if a key type says it's `ConfigurableString` but actually generates a non-string-like schema.